pkg/providers/tinkerbell/validate.go

```diff
@@ -162,14 +162,82 @@ func validateK8sVersionInOSImageURLs(spec *ClusterSpec) error {
 			}
 
 			if !containsK8sVersion(url, string(version)) {
-				return fmt.Errorf("missing kube version from worker node group machine config OSImageURL: url=%v, version=%v",
-					url, version)
+				return fmt.Errorf("missing kube version from worker node group machine config OSImageURL, the image url must include %v: url=%v, version=%v",
+					permutations(string(version), []string{"-", "_", ""}), url, version)
 			}
 		}
 	}
 	return nil
 }
 
+// permutations takes a version in dot format and
+// returns a human readable string with the permutations of the version
+// using the passed in separators.
+//
+// For example, when v = 1.29 and separators = []string{"-", "_", ""}
+// the result will be "1.29, 1-29, 1_29, or 129".
+func permutations(v string, separators []string) string {
+	result := []string{v}
+	split := strings.Split(v, ".")
+	if len(split) == 1 {
+		return v
+	}
+
+	seps := remove(".", deduplicate(separators))
+	for idx, sep := range seps {
```

Comment:
can we simplify this logic to just parsing in the version and concatenating the seps to list out all accepted versions in a single for loop
```suggestion
        version := strings.Split(v, ".")
        result := []string{v}
        major := version[0]
        minor := version[1]
	for idx, sep := range seps {
	       elem := fmt.Sprintf("%s%s%s",major,sep,minor)

	       if idx == len(seps)-1 {
	       result = append(result, "or")
	       } else if idx != len(seps)-2 {
			elem = elem + ","
		}
		result = append(result, elem)
	}
	return strings.Join(result, " ")
}
```

Reply:
this changes the function to only support a single `.` as a separator. In my opinion, this more generic implementation is basic, flexible, and simple enough to be future proof. I'm not convinced of the value of the trade off of a few less lines of code.